### DISCUSSION TITLE
No morph map defined for model [App\Models\User]

### DISCUSSION THREAD

**Discussion by ClemtM** [spatie/laravel-permission#2148] [Q&A] [answered]
Hi,

I have troubles when I try to install spatie/laravel-permission. I have the error **No morph map defined for model [App\Models\User]**

When I installed spatie with _composer require spatie/laravel-permission_, I got that :

> @php artisan ide-helper:eloquent
Unexpected no document on Illuminate\Database\Eloquent\Model
Wrote expected docblock to C:\Users\Clement\Documents\Projets\Laravel\BDV\bdv-laravel-alex\vendor\laravel\framework\src\Illuminate\Database\Eloquent\Model.php
> @php artisan ide-helper:generate
A new helper file was written to _ide_helper.php
> @php artisan ide-helper:meta
A new meta file was written to .phpstorm.meta.php
> @php artisan ide-helper:models --nowrite
Error resolving relation model of App\Models\User:notifications() : No morph map defined for model [App\Models\User].        
Error resolving relation model of App\Models\User:permissions() : No morph map defined for model [App\Models\User].
Error resolving relation model of App\Models\User:permissions() : No morph map defined for model [App\Models\User].
Error resolving relation model of App\Models\User:roles() : No morph map defined for model [App\Models\User].
Error resolving relation model of App\Models\User:roles() : No morph map defined for model [App\Models\User].
Model information was written to _ide_helper_models.php

I also updated composer, check my User model, recheck the doc but still.
And when I tried to use package fonctions from user, I get the same issue.
I try again to install it but nothing changed. 

I think there is someting with either my installation, my requirements or my application itself. Do you have any idea ?

The User model : https://pastebin.com/rp05XFaA

![image](https://user-images.githubusercontent.com/105287888/182566033-00f49375-9187-49fd-990c-9cb6888fedda.png)

Edit : I update Laravel 8 to 9, but it doent solve my issue

**Comment by ClemtM** [comment, marked as answer]
Ok I got it. The issue was because I used Relation::enforceMorphMap in the AppServiceProvider.

Maybe it will help to know that

**Reply by sergiuastechsoft** [reply]
Nice :D good job

**Reply by Sjoerd24** [reply]
In case anyone else comes across this and wants to keep using morphmaps (with spatie permissions), you have to add the following to your list:

       Relation::enforceMorphMap([
            // your other models and these two:
            'permission' => 'Spatie\Permission\Models\Permission',
            'role' => 'Spatie\Permission\Models\Role',
            
        ]);

**Comment by uyson** [comment]
How can I fix this exception

**Reply by angeljqv** [reply]
https://github.com/spatie/laravel-permission/discussions/2148#discussioncomment-3315498

**Reply by uyson** [reply]
Why I can't use  Relation::enforceMorphMap ? I need this

**Reply by angeljqv** [reply]
>Why I can't use Relation::enforceMorphMap ?

you can use it, nobody says otherwise, I'm using it without any problem, the correct question would be "What are you doing wrong in your implementation?", and that question can only be answered by you

**Reply by quyle92** [reply]
> Why I can't use Relation::enforceMorphMap ? I need this

same issue here... After reading through the doc, I figured it out:
Using  `Relation::enforceMorphMap `means, it will toggle the `$requireMorphMap` property on:
```
public static function requireMorphMap($requireMorphMap = true)
    {
        static::$requireMorphMap = $requireMorphMap;
    }
```
then whenever you called the morph relation of User instance, it will call `getMorphClass()` where it will check whether `$requireMorphMap` is true via `Relation::requiresMorphMap()`, if it does, exception will be thrown.
```
 public function getMorphClass()
    {
       ...
        if (Relation::requiresMorphMap()) {
            throw new ClassMorphViolationException($this);
        }

        return static::class;
    }
```

Solution: 
1/ Add a new key to your Relation::enforceMorphMap() with the model that is causing the error.
2/ Replace the call to Relation::enforceMorphMap() with Relation::morphMap(). This will not require the morph maps anymore for your models.

Inspired by : [https://ralphjsmit.com/laravel-fix-no-morph-map-defined](https://ralphjsmit.com/laravel-fix-no-morph-map-defined)
